### PR TITLE
FB-214: Site as a model

### PR DESCRIPTION
**Pull request by fvovan**
It's the second option of https://github.com/cargomedia/cm/issues/2505 solution

### CODE REVIEW

File: library/CM/FormField/Site.php [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 class CM_FormField_Site extends CM_FormField_Set_Select {
 
+    //TODO fix method
```
Comment (njam) [inline library/CM/FormField/Site.php]: todo

**Comment by fvovan** [comment]
It's more or less draft again (i mean going further with sk changes most likely will require to adjust something here)
first part of SK changes is here https://upsource.cargomedia.ch/sk/review/SK-CR-383 (not even close to be finished)

As  for me, maybe i would prefer separate models for site settings as i did in https://upsource.cargomedia.ch/cm/review/CM-CR-11. Site models concept itself is not really clear   since (as we  discussed) it will be anyway impossible to create new site without any application release.
and besides there are some inconveniences.

1) We'll lose the ability to refer to site objects with their constructor calls. So each `new FB_Site_Fuckbook_Cam()` will be changed to something like `(new CM_Site_Factory())->getSiteByType(FB_Site_Fuckbook_Cam::getTypeStatic())`. I.e. we need either siteID (and it's not clear how to obtain it in general case), or `siteType`.
2) we'll have to migrate 9 mysql tables and 1 mongo collection to contain site id. Though it could be `siteType` as it's now (since we agreed `type` will be unique)

And it the end we get back only database schemas which probably can be implemented in separate from `CM_Site_Abstract` classes.

**Comment by fvovan** [comment]
As discussed with @{c7e8ab9c-90bf-4796-bb88-9aa7c8150044,Reto Kaiser},
1) use `FB_Site_Fuckbook_Cam::factory()` instead of constructors
2) keep db for now
3) implement admin part of settings later
4) for sk part move to db only basic settings (`name` and `email`)
5) look into example data scripts later when sk part is near to be finished

File: library/CM/Model/Schema/Definition.php [hunk 2]
```diff
@@ -124,7 +124,12 @@ public function decodeField($key, $value) {
                                 throw new CM_Model_Exception_Validation('Field type is not a valid class/interface', null, ['type' => $type]);
                             }
                             $className = $type;
-                            if (is_a($className, 'CM_Model_Abstract', true)) {
+                            if (is_a($className, CM_Site_Abstract::class, true)) {
+                            /** @var CM_Site_Abstract $type */
+                                $id = $value;//TODO check if we need JSON encoded values
+                                $value = CM_Site_Abstract::factoryFromId($id);
+                            }
+                            elseif (is_a($className, 'CM_Model_Abstract', true)) {
```
Comment (njam) [inline library/CM/Model/Schema/Definition.php]: Formatting

File: library/CM/Model/Schema/Definition.php [hunk 3]
```diff
@@ -124,7 +124,12 @@ public function decodeField($key, $value) {
                                 throw new CM_Model_Exception_Validation('Field type is not a valid class/interface', null, ['type' => $type]);
                             }
                             $className = $type;
-                            if (is_a($className, 'CM_Model_Abstract', true)) {
+                            if (is_a($className, CM_Site_Abstract::class, true)) {
+                            /** @var CM_Site_Abstract $type */
+                                $id = $value;//TODO check if we need JSON encoded values
```
Comment (njam) [inline library/CM/Model/Schema/Definition.php]: what about this todo?

File: library/CM/Site/Abstract.php [hunk 4]
```diff
@@ -2,7 +2,7 @@
```
Comment (njam) [inline library/CM/Site/Abstract.php]: |`line 357`|
|---|
Should we change it to `cm_site`?

**Comment by njam** [comment]
|`library/CM/Site/Abstract.php:119`|
|---|
In this case I think it's better to invalidate all relevant caches in `setDefault()`. In the end we just need to invalidate the cache of thee *previous* default site, and of the *new* default site. Or if it's easier we can invalidate the caches of *all* sites.

**Comment by njam** [comment]
|`library/CM/Site/Abstract.php:132`|
|---|
Alternatively how about getting the *previous* default site instance, ans setting `setDefault(false)` on it?
On the other hand, it would be nice if this was an atomic operation. Can it be done in a single mongodb query?

**Comment by njam** [comment]
|`library/CM/Site/Abstract.php:119`|
|---|
In this case I think it's better to invalidate all relevant caches in `setDefault()`. In the end we just need to invalidate the cache of thee *previous* default site, and of the *new* default site. Or if it's easier we can invalidate the caches of *all* sites.

**Comment by njam** [comment]
|`library/CM/Site/Abstract.php:281`|
|---|
Use getUrlString() for better performance

**Comment by njam** [comment]
|`library/CM/Site/Abstract.php:337`|
|---|
Should we add caching for this?

**Comment by njam** [comment]
|`library/CM/Site/Abstract.php:372`|
|---|
Let's replace it with a proper `public static function create($proper, $signature)`?

**Comment by fvovan** [comment]
Not sure it's needed. We are not going to call `CM_Class_Abstract::create()` anyway. 
`_createStatic()` is for `SK_SIte_Abstract::create()` (and its descendants usages)

**Comment by njam** [comment]
agree

**Comment by fvovan** [comment]
may be but retrieval by ID should be very fast, as it stores indexes in RAM. but i need to check how it works with our new mongo

**Comment by njam** [comment]
Still it's magnitudes slower than doing a lookup in memory.
Since this function is used when decoding models, we really need to assume it's used a lot - because we don't know ''when'' it's used.
I assume the "_type" of a site cannot be changed? If so then cache invalidation is not an issue, and we can simply cache this in "CacheLocal" - easy and impactful.

**Comment by njam** [comment]
Still it's magnitudes slower than doing a lookup in memory.
Since this function is used when decoding models, we really need to assume it's used a lot - because we don't know ''when'' it's used.
I assume the "_type" of a site cannot be changed? If so then cache invalidation is not an issue, and we can simply cache this in "CacheLocal" - easy and impactful.

**Comment by fvovan** [comment]
okay, but why not use `$id` as cache key?

**Comment by njam** [comment]
Yes $id as cache key sounds good.

**Comment by fvovan** [comment]
Currently it's not possible to *unset* the field (we need this to allow additional index level guard, setting value to `null` doesn't work btw).
MongoDB doesn't support atomicity across several documents update. How about just resetting the cache?

**Comment by njam** [comment]
sgtm
btw, please don't "resolve" my comments - I'd like to use it for checking it once more, thanks

**Comment by fvovan** [comment]
ah. sorry. I thought it's a part of the upsource workflow

**Comment by njam** [comment]
|`library/CM/Site/Abstract.php:120`|
|---|
How about just making this not optional? Seems it's mandatory.

**Comment by njam** [comment]
|`library/CM/Site/Abstract.php:128`|
|---|
If we pass FALSE we can end up with no default anymore, which will be a problem. How about some sanity checks?

**Comment by fvovan** [comment]
It can't be mandatory. The thing it's indexed as *unique* key in the collection, so `true` value may exist only once. The same is for `false` and `null`

**Comment by fvovan** [comment]
I removed method parameter.

**Pull request merged**